Correctly reduce with min/max over Views full of +inf/-inf

PaulGannay opened the pull request:
Currently, when reducing using the `Min` reducer over a View containing only +inf, we get FLOAT_MAX as a result ([reproducer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:30,positionColumn:2,positionLineNumber:30,selectionStartColumn:2,selectionStartLineNumber:30,startColumn:2,startLineNumber:30),source:'%23include+%3Ciostream%3E%0A%23include+%3Climits%3E%0A%0A%23include+%3CKokkos_Core.hpp%3E%0A%0Aint+main(int+argc,+char*+argv%5B%5D)+%7B%0A++++Kokkos::ScopeGuard+guard(argc,+argv)%3B%0A++++%7B%0A++++++++const+int+N+%3D+10%3B%0A++++++++%0A++++++++Kokkos::View%3Cfloat*%3E+f(%22view%22,+N)%3B%0A++++++++Kokkos::parallel_for(N,+KOKKOS_LAMBDA(const+int+i)%7B%0A++++++++++++f(i)+%3D+std::numeric_limits%3Cfloat%3E::infinity()%3B%0A++++++++%7D)%3B%0A++++++++%0A++++++++float+min%3B%0A++++++++Kokkos::parallel_reduce(N,+KOKKOS_LAMBDA+(const+int%26+i,+float%26+res)+%7B%0A++++++++++++res+%3D+res+%3C+f(i)+%3F+res+:+f(i)%3B%0A++++++++%7D,+Kokkos::Min%3Cfloat%3E(min))%3B%0A++++++++%0A++++++++std::cout+%3C%3C+min+%3C%3C+std::endl%3B%0A%0A++++++++min+%3D+std::numeric_limits%3Cfloat%3E::infinity()%3B%0A++++++++for+(int+i+%3D+0%3B+i%3CN%3B+%2B%2Bi)+%7B%0A++++++++++++min+%3D+min+%3C+f(i)+%3F+min+:+f(i)%3B%0A++++++++%7D%0A++++++++std::cout+%3C%3C+min+%3C%3C+std::endl%3B%0A%0A++++%7D%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:53.43450479233227,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g121,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:kokkos,ver:'4500')),options:'-std%3Dc%2B%2B23+-Wall+-Wextra+-Werror',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+12.1+(Editor+%231)',t:'0')),k:50,l:'4',m:50.32010243277849,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+12.1+(Compiler+%231)',t:'0')),header:(),l:'4',m:49.67989756722151,n:'0',o:'',s:0,t:'0')),k:46.56549520766773,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)).
With this PR, we should correctly return +inf/-inf.

PaulGannay committed:
Correctly reduce with min/max over Views full of +inf/-inf

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,65 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+
+TEST(TEST_CATEGORY, longdouble_reduction_over_infinite) {
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  // This won't compile if TEST_EXECSPACE is not on host because long double =
+  // double on device.
+  // KOKKOS_IF_ON_HOST(TestReductionOverInfiniteFloat<long double>();)
+#endif
```
Not sure how to correctly detect if `TEST_EXECSPACE` is Serial/OpenMP/...

masterleinad replied:
You could ask if `TEST_EXECSPACE::memory_space` is `HostSpace` for example.

masterleinad commented on core/src/Kokkos_ReductionIdentity.hpp:
```diff
@@ -363,8 +366,17 @@ struct reduction_identity<float> {
   KOKKOS_FORCEINLINE_FUNCTION constexpr static float prod() {
     return static_cast<float>(1.0f);
   }
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_4
```
In my mind it's fixing a bug so we don't need to deprecate.

dalg24 replied:
Agreed.
We will see if it breaks Trilinos in the nightlies.

dalg24 commented on core/src/Kokkos_ReductionIdentity.hpp:
```diff
@@ -363,8 +366,17 @@ struct reduction_identity<float> {
+  }
+#else
   KOKKOS_FORCEINLINE_FUNCTION constexpr static float max() { return -FLT_MAX; }
   KOKKOS_FORCEINLINE_FUNCTION constexpr static float min() { return FLT_MAX; }
```
For consistency I would use `HUGE_VALF`

dalg24 commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,65 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+
+// We do not run the test if KOKKOS_ENABLE_DEPRECATED_CODE_4 is ON because the
+// fix is only available if it is OFF
+TEST(TEST_CATEGORY, float_reduction_over_infinite) {
```
```suggestion
TEST(TEST_CATEGORY, reduction_identity_min_max_floating_point_types) {
```
Instantiate all 3 here and inject `<< Kokkos::Impl::TypeInfo<T>::name() << '\n';` in the assertions

dalg24 reviewed (commented):
Thanks for fixing this.

Adrien-Tab commented:
I'm pretty sure this problem also exists with half type

PaulGannay committed:
Apply review comment

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,56 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+  TestReductionOverInfiniteFloat<Kokkos::Experimental::bhalf_t>();
+  TestReductionOverInfiniteFloat<float>();
+  TestReductionOverInfiniteFloat<double>();
+  if constexpr (std::is_same_v<TEST_EXECSPACE::memory_space,
+                               Kokkos::HostSpace>) {
+    // TestReductionOverInfiniteFloat<long double>();
+  }
```
This is not enough to stop nvcc from compiling `runTest()` with `long double` for the device.
I tried to add `std::enable_if` but couldn't make it work.

masterleinad replied:
Then check for the default memory space instead.

JBludau replied:
afaik `if constexpr` will still require all paths to be compilable. I think you will have to #ifdef or sfinae (which does not make any sense here)

dalg24 replied:
I expect you meant paths -> passes in reference to NVCC's two stage compilation model?

dalg24 replied:
Or did you mean device and host paths?

PaulGannay replied:
I misread the warning I had, the problem is not that I couldn't stop nvcc from compiling the reduction for `long double`, it is that we can't write a reduction with `long double` without generating warnings with nvcc. 
Even the following reduction generates warnings: 
```C++
    void test() { 
      long double res;
      Kokkos::parallel_reduce(
          Kokkos::RangePolicy<Kokkos::DefaultHostExecutionSpace>(0,0),
          KOKKOS_LAMBDA(const int i, long double& partial_res){},
          Kokkos::Min<long double>(res)); 
    }
```

So now the question is: should the correction of warnings when using `long double` should be part of this PR (with the risk that it may not be ready for inclusion in 4.7), or should I correct the warning in a new PR and totally disable the tests with `long double`?

PaulGannay replied:
And what may makes the warning somewhat slow to remove is that nvcc doesn't give a reference to a file or line when emitting it, so finding every instance where we need to add the `pragma diag_supress` may take some time.

For the record, full warning is:
```
[15/19] Building CXX object core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Develop.dir/default/TestDefaultDeviceDevelop.cpp.o
Warning #20208-D: 'long double' is treated as 'double' in device code

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"
```

dalg24 replied:
Do you get that warning earn compiling with Clang(+CUDA)?  with HIP?  with SYCL?

dalg24 replied:
In the math function tests we `#ifdef` away `long double` overloads testing if any of the device backends is enabled.  In the numeric traits, we use pragma directives to suppress the warnings with NVCC.
Another trip would be to do `if constexpr (is_same_v<DefaultExecutionSpace, DefaultHostExecutionSpace>)`

PaulGannay replied:
> In the math function tests we `#ifdef` away `long double` overloads testing if any of the device backends is enabled. In the numeric traits, we use pragma directives to suppress the warnings with NVCC.

It is a bit ugly, but it works when using both at the same time.
The problem I had with the pragma directive is that you have to put them before the `Kokkos_Core.hpp` include.

Thank you for your help.

science-enthusiast replied:
> I expect you meant paths -> passes in reference to NVCC's two stage compilation model?

I think @JBludau was referring to the code path that a true if condition would take. In this case it is `TestReductionOverInfiniteFloat<long double>();`. Suppose `TestReductionOverInfiniteFloat<long double>();` is not compilable for a given backend then even if the condition in `if constexpr` can avoid that code path, it is not acceptable for the C++ compiler.

PaulGannay commented:
> I'm pretty sure this problem also exists with half type

I added test for half types.

masterleinad commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,56 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+          }
+        },
+        Kokkos::Max<ScalarType>(min));
+    ASSERT_EQ(-inf, min) << "For type "
```
```suggestion
        Kokkos::Max<ScalarType>(max));
    ASSERT_EQ(-inf, max) << "For type "
```

dalg24 committed:
Fixup copy/pasta error min -> max

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

dalg24 committed:
Fixup add missing <cmath> header include

Signed-off-by: Damien L-G <dalg24@gmail.com>
Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

science-enthusiast commented:
What are the challenges in initializing the minimum/maximum value to the first element of the array of values and then performing the reduction? This is as suggested by the comment by Michael Madsen in this [SO thread](https://stackoverflow.com/questions/4675601/how-do-arbitrary-floating-point-values-compare-to-infinity). The reduction will also work if all the values are +infinity or -infinity.

dalg24 commented:
> What are the challenges in initializing the minimum/maximum value to the first element of the array of values and then performing the reduction? This is as suggested by the comment by Michael Madsen in this [SO thread](https://stackoverflow.com/questions/4675601/how-do-arbitrary-floating-point-values-compare-to-infinity). The reduction will also work if all the values are +infinity or -infinity.

`NaN`, memory accessibility, parallelization

science-enthusiast commented:
> `NaN`, memory accessibility, parallelization

`NaN` especially is a critial problem.

science-enthusiast commented on core/src/Kokkos_ReductionIdentity.hpp:
```diff
@@ -363,8 +364,10 @@ struct reduction_identity<float> {
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static float max() {
+    return -HUGE_VALF;
+  }
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static float min() { return HUGE_VALF; }
```
So far, `max()` (resp. `min()`) have meant the minimum (resp. maximum) representable value of that type. Now, it will not mean that for the floating point types. `HUGE_VAL` etc are usually encountered in the case of [overflow](https://en.cppreference.com/w/cpp/numeric/math/HUGE_VAL).

At the same time, these values are grouped under `reduction_identity` for each type. That gives better context. Still it could be a surprise for the user. Maybe, we could document it.

PaulGannay replied:
The name are a bit confusing but `reduction_identity<T>::min` doesn't designate the `min` value of `T` but the value used to initialised the `Min<T>` reducer.

science-enthusiast replied:
Yes, `reduction_identity` clarifies the context to the user. At the same time, we could document this behavior for floating point numbers.

PaulGannay replied:
This is not specific to floating point numbers, for all types `T`,  `reduction_identity<T>::min` returns `max<t>`.

dalg24 replied:
Take this discussion to https://github.com/kokkos/kokkos-core-wiki/issues/163

science-enthusiast commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,56 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+  TestReductionOverInfiniteFloat() { runTest(); }
+
+  void runTest() {
+    const int N = 10;
```
This could be an unsigned int. `10` is a small value only but it will be consistent with many of the other tests in `core/unit_test/TestReduce.hpp`, where View size, work count etc are unsigned int.

PaulGannay committed:
Remove warning in TestReduce

Do not compile for `long double` when a GPU backend is active.
Disable the "'long double' is treated as 'double' in device code" warning when compiling with nvcc.

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

science-enthusiast commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,56 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+  void runTest() {
+    const int N = 10;
+
+    ScalarType inf = std::numeric_limits<ScalarType>::infinity();
```
`Kokkos::Experimental::infinity_v` from `core/src/Kokkos_NumericTraits.hpp` could be used. Many other parts of Kokkos are using that.

PaulGannay replied:
We don't need infinity to be available on the device, so no need to depend on an Experimental feature.

dalg24 replied:
numeric_limits is fine

science-enthusiast replied:
Your point is noted. `Kokkos::Experimental::infinity<T>` is ultimately defined as `std::numeric_limits<T>::infinity()` in `core/src/Kokkos_NumericTraits.hpp`. I do see `std::numeric_limits<T>::max()` etc being called in other unit_tests. So calling `std::numeric_limits<T>::infinity()` should be fine.

dalg24 replied:
It is actually a bug because we get 0 for the half types.
We need to reaudit if the test is sturdy enough.

PaulGannay replied:
Indeed, it is 0.
I modified the test to use `Kokkos::Experimental::infinity_v`, which returns the correct value for half and bhalf types.

science-enthusiast replied:
When you say correct value, I suppose you mean infinity for those types. It will be nice if someone can say where in Kokkos are the specializations of `std::is_floating_point_v<T>` and `std::numeric_limits<T>::infinity()` for `KE::half_t` and `KE::bhalf_t`. Which headers to include in order to access these specializations?

Maybe, it is `Kokkos::Experimental::infinity<T>` that is specialized for these types. I see  `Kokkos::Experimental::Impl::infinity_helper` in `core/src/impl/Kokkos_Half_NumericTraits.hpp` for the two types~~but I am not yet able to find how they are used~~. I was in a hurry. I see how `infinity_helper` of the two types are used inside the macro in `core/src/Kokkos_NumericTraits.hpp`.

PaulGannay replied:
> `std::numeric_limits<T>::infinity()`

It is in `core/src/impl/Kokkos_Half_NumericTraits.hpp`.

science-enthusiast commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +672,56 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+ public:
+  TestReductionOverInfiniteFloat() { runTest(); }
+
+  void runTest() {
```
`runTest()` could be private.

dalg24 replied:
Don't do that, it is not supported by NVCC
https://docs.nvidia.com/cuda/cuda-c-programming-guide/#extended-lambda-restrictions (clause 4)

PaulGannay committed:
Apply PR comments

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

dalg24 commented on core/unit_test/TestReduce.hpp:
```diff
@@ -18,6 +18,20 @@
+#pragma diag_suppress 3245
+#endif
+#endif
+#endif
```
Why both the suppression and ifdefing out the long double instantiation?

PaulGannay replied:
You are right, I got confused, the suppression is the only one necessary.

dalg24 commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +686,68 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+};
+
+TEST(TEST_CATEGORY, reduction_identity_min_max_floating_point_types) {
+  TestReductionOverInfiniteFloat<Kokkos::Experimental::half_t>();
+  TestReductionOverInfiniteFloat<Kokkos::Experimental::bhalf_t>();
```
BTW I expected these would fail with CUDA, HIP, and SYCL...

PaulGannay replied:
I was also expecting it to fail and was surprised to see it didn't.

Adrien-Tab replied:
Is `std::numeric_limits<T>` defined with half type.

science-enthusiast replied:
> Is `std::numeric_limits<T>` defined with half type.

[cppreference](https://en.cppreference.com/w/cpp/types/numeric_limits.html) says that it depends on the implementation. E.g., OpenEXR provides a `std::numeric_limits<half>` specialization, it seems.

dalg24 replied:
Holly shit it will give zero...

PaulGannay committed:
Remove useles pragma that were disabling some warnings

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay committed:
Correctly initialized reduce operation over half and bhalf

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

dalg24 commented on core/unit_test/TestReduce.hpp:
```diff
@@ -18,6 +18,8 @@
 #include <limits>
 
+#include <Kokkos_Macros.hpp>
+
```
```suggestion
```

dalg24 commented on core/unit_test/TestReduce.hpp:
```diff
@@ -672,4 +674,59 @@ TEST(TEST_CATEGORY, reduction_with_large_iteration_count) {
+  void runTest() {
+    const unsigned int N = 10;
+
+    ScalarType inf = Kokkos::Experimental::infinity_v<ScalarType>;
```
We need to ensure that `inf` is what we think it is to avoid the embarrassment we had when using std::numeric_limits.
Something like abort or increment an error counter if `inf!=inf+1`

PaulGannay replied:
You are right, but I used `inf==inf*inf` because `inf==inf+1` can't distinguish between `FLOAT_MAX` and `HUGE_VAL`.

dalg24 replied:
maybe inf==inf+1 && inf=inf-1
because one*one==one

PaulGannay replied:
You are right, it wouldn't detect if inf is 1 or 0.
I added the check for `inf == inf+1`, it should be enough.

science-enthusiast replied:
An interesting discussion related to this: https://stackoverflow.com/questions/58357674/smallest-number-added-to-flt-max-to-cause-overflow

dalg24 commented on core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp:
```diff
@@ -484,10 +484,10 @@ struct reduction_identity<Kokkos::Experimental::bhalf_t> {
   }
   KOKKOS_FORCEINLINE_FUNCTION constexpr static float max() noexcept {
-    return -0x7f7f;
+    return -Kokkos::Experimental::infinity_v<float>;
```
~~Why float?~~  OK I see the return type

Also we need to think carefully about not adding another circular dependency

science-enthusiast replied:
We could also check whether the casting works as expected for float to half_t and bhalf_t for the special values infinity and -infinity.

dalg24 replied:
Sure but that sounds like another PR

PaulGannay committed:
Adds check to ensure that infinity is infinity

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay committed:
Correctly detect that infinity is infinity

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

dalg24 commented on core/src/SYCL/Kokkos_SYCL_Half_Conversion.hpp:
```diff
@@ -116,12 +116,11 @@ struct reduction_identity<Kokkos::Experimental::half_t> {
   max() noexcept {
-    return std::numeric_limits<
-        Kokkos::Experimental::half_t::impl_type>::lowest();
+    return -Kokkos::Experimental::infinity_v<Kokkos::Experimental::half_t>;
```
If we use the numeric traits then we need to make sure that the half type specialization is visible from here.

crtrott commented:
Do we have any tests where say every value expect one is inf, and we get back the right result?

PaulGannay commented:
> Do we have any tests where say every value expect one is inf, and we get back the right result?

As far as I know, no (it is not part of this PR anyway).
I don't think it should ever fail (inf is not special when doing a comparison?) but if you think it might be worth to test I can add one.

PaulGannay commented:
Regarding the nvhpc test failure, it is a bug in nvhpc itself.
The following code incorrectly returns `3.40282e+38` when executed on GPU, and correctly returns `inf` when executed on CPU:
```C++
#include <iostream>
#include <cmath>

int main(){

  const int N = 10;
  
  float array[N];
  for (int i=0; i<N; ++i) {
    array[i] = HUGE_VALF;
  }

  float min = HUGE_VALF;

#pragma acc parallel loop vector reduction(min:min) copyin(array)
    for (auto i = 0; i < N; i++) {
      if (array[i] < min) min = array[i];
     }

    std::cout << min << std::endl;
}
```

For now I guess that the best course of action would be to disable the test specifically for nvhpc?
I'll report the bug to nvidia.

PaulGannay committed:
Adds missing include to make the half type definition available when using traits from Kokkos_NumericTraits

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay committed:
Revert "Adds missing include to make the half type definition available when using traits from Kokkos_NumericTraits"

This doesn't work on backends where half_t and bhalf_t are not available

This reverts commit 92380ae5cc194623c23f329ea1e2ceb9f560428e.

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

PaulGannay committed:
Disable tests for nvhpc

Signed-off-by: Paul Gannay <paul.gannay@cea.fr>

masterleinad committed:
Fix SYCL

Signed-off-by: Daniel Arndt <arndtd@ornl.gov>

masterleinad reviewed (approved):
I fixed the SYCL build. The rest looks good to me.

JBludau reviewed (commented):
have we tried what happens in the case of ?

[GCC](https://gcc.gnu.org/wiki/FloatingPointMath) and [Clang](https://clang.llvm.org/docs/UsersManual.html#controlling-floating-point-behavior) support a -ffinite-math option (additionally implied by -ffast-math), which allows the respective compiler to assume the nonexistence of special IEEE-754 floating point values such as NaN, infinity, or negative zero.

crtrott commented:
> have we tried what happens in the case of ?
> 
> 
> [GCC](https://gcc.gnu.org/wiki/FloatingPointMath) and [Clang](https://clang.llvm.org/docs/UsersManual.html#controlling-floating-point-behavior) support a -ffinite-math option (additionally implied by -ffast-math), which allows the respective compiler to assume the nonexistence of special IEEE-754 floating point values such as NaN, infinity, or negative zero.
> 

Good question.

crtrott commented:
This test passes with fast-math but these don't:
```
[  FAILED  ] 10 tests, listed below:
[  FAILED  ] serial.atomic_operations_complexfloat
[  FAILED  ] serial.atomic_operations_double
[  FAILED  ] serial.atomic_operations_float
[  FAILED  ] serial.mathematical_functions_absolute_value
[  FAILED  ] serial.mathematical_functions_floating_point_absolute_value
[  FAILED  ] serial.mathematical_functions_remainder_function
[  FAILED  ] serial.mathematical_functions_isfinite
[  FAILED  ] serial.mathematical_functions_isinf
[  FAILED  ] serial.mathematical_functions_isnan
[  FAILED  ] serial.mathspecialfunc_errorfunc
```
Serial build on my Mac. They also fail on develop branch.

science-enthusiast commented on core/src/SYCL/Kokkos_SYCL_Half_Conversion.hpp:
```diff
@@ -114,12 +115,16 @@ struct reduction_identity<Kokkos::Experimental::half_t> {
   }
-  KOKKOS_FORCEINLINE_FUNCTION constexpr static Kokkos::Experimental::half_t
-  min() noexcept {
+  KOKKOS_FORCEINLINE_FUNCTION constexpr static auto min() noexcept {
```
@masterleinad Please take your time to clarify these doubts.

After this commit, in `min` and `max` return type is being inferred. How is it related to sycl::half not having constexpr constructors? It looks like there is a relationship because for `min` that is the only change.

Also, in `max` why does an explicit `bit_comparison_type` based value help to signify -Infinity? Infinity also corresponds to a `bit_comparison_type` based value in the case of `half_t`.

masterleinad replied:
There is no `constexpr` conversion between `bit_comparison_type` and `half_t` since our `bit_cast` implementation isn't (and can't be) `constexpr`. For that we need compiler support.

`bit_comparison_type` doesn't have a unitary minus operator allowing to flip the sign. Hence, we need to construct the bit representation here.

science-enthusiast replied:
Thank you, @masterleinad . It is a lot clearer now.
